colmena/proxy.py

```diff
@@ -1,85 +1,58 @@
-    """Custom Factory for using Colmena serialization utilities"""
-    def __init__(self,
-                 key: str,
-                 name: str,
-                 hostname: str,
-                 port: int,
-                 serialization_method: Union[str, SerializationMethod] = SerializationMethod.PICKLE,
-                 **kwargs) -> None:
-        """Init ColmenaSerialization Factory
-
-        Args:
-            key (str): key corresponding to object in Redis.
-            name (str): name of store to retrive objects from.
-            hostname (str): hostname of Redis server containing object.
-            port (int): port of Redis server containing object.
-            serialization_method (str): Colmena serialization method to use
-                for deserializing the object when resolved from Redis.
-            kwargs: keyword arguments to pass to the RedisFactory.
-        """
-        self.serialization_method = serialization_method
-        self.kwargs = kwargs
-        super(ColmenaSerializationFactory, self).__init__(
-            key, name, hostname, port, **kwargs
-        )
-
-    def __getnewargs_ex__(self):
-        """Helper method for pickling
-
-        Note:
-            We override default pickling behavior because a Factory may contain
-            a Future if it is being asynchronously resolved and Futures cannot
-            be pickled.
-        """
-        return (self.key, self.name, self.hostname, self.port), {
-            'serialization_method': self.serialization_method,
-            **self.kwargs
-        }
-
-    def resolve(self) -> Any:
-        obj_str = super(ColmenaSerializationFactory, self).resolve()
-        return SerializationMethod.deserialize(self.serialization_method, obj_str)
-
-
-def proxy(obj: Any,
-          key: Optional[str] = None,
-          is_serialized: bool = False,
-          serialization_method: Union[str, SerializationMethod] = SerializationMethod.PICKLE,
-          **kwargs) -> ps.proxy.Proxy:
-    """Place object in Value Server and return Proxy
+from typing import Any, Union
+
+
+class ProxyJSONSerializationWarning(Warning):
+    pass
+
+
+def proxy_json_encoder(proxy: ps.proxy.Proxy) -> Any:
+    """Custom encoder function for proxies
+
+    Proxy objects are not JSON serializable so this function, when passed to
+    `json.dumps()`, will attempt to JSON serialize the wrapped object. If the
+    proxy is not resolved, a warning will be raised for the user and the
+    proxy will be replaced with a placeholder string for the proxy. This
+    1) prevents JSON serialization from failing and 2) avoid unintended
+    resolutions of proxies that may invoke expensive communication operations
+    without the user being aware.
+
+    Usage:
+        >>> # With JSON dump/dumps
+        >>> json.dumps(json_obj_containing_proxy, default=proxy_json_encoder)
+        >>> # With Pydantic
+        >>> my_basemodel_instance.json(encoder=proxy_json_encoder)
```

Comment:
Does this mean we could use `proxy_json_encoder` for all fields? 

If so, we can code that as the default for `Result`: https://pydantic-docs.helpmanual.io/usage/exporting_models/#custom-json-deserialisation

Reply:
That was my first though as well, but all of the proxies were still being resolved. I think it has something to do with how pydantic recursively inspects the elements inside the object inadvertently resolving the proxies.